cpp/src/arrow/compute/kernels/vector_is_monotonic.cc

```diff
@@ -0,0 +1,317 @@
+    *strictly_increasing = false;
+    *decreasing = false;
+    *strictly_decreasing = false;
+  } else {
+    bool equal =
+        // Approximately equal within some error bound (epsilon).
+        (options.floating_approximate &&
+         (fabs(current - next) <=
+          static_cast<typename DataType::c_type>(options.epsilon))) ||
+        // Or exactly equal.
+        current == next;
+    if (*increasing) {
+      if (!(equal || next > current)) {
+        *increasing = false;
+        *strictly_increasing = false;
+      }
+    }
+    if (*decreasing) {
+      if (!(equal || next < current)) {
+        *decreasing = false;
+        *strictly_decreasing = false;
+      }
+    }
+    if (*strictly_increasing) {
+      if (equal || !(next > current)) {
+        *strictly_increasing = false;
+      }
+    }
+    if (*strictly_decreasing) {
+      if (equal || !(next < current)) {
+        *strictly_decreasing = false;
+      }
+    }
+  }
+}
+
+template <typename DataType>
+enable_if_not_floating_point<DataType> IsMonotonicCheck(
+    const typename DataType::c_type& current, const typename DataType::c_type& next,
+    bool* increasing, bool* strictly_increasing, bool* decreasing,
+    bool* strictly_decreasing, const IsMonotonicOptions& options) {
+  if (*increasing) {
+    if (!(next >= current)) {
+      *increasing = false;
+      *strictly_increasing = false;
+    }
+  }
+  if (*strictly_increasing) {
+    if (!(next > current)) {
+      *strictly_increasing = false;
+    }
+  }
+  if (*decreasing) {
+    if (!(next <= current)) {
+      *decreasing = false;
+      *strictly_decreasing = false;
+    }
+  }
+  if (*strictly_decreasing) {
+    if (!(next < current)) {
+      *strictly_decreasing = false;
+    }
+  }
+}
+
+template <typename DataType>
+enable_if_floating_point<DataType, bool> isnan(
+    const util::optional<typename DataType::c_type>& opt) {
+  return opt.has_value() && std::isnan(opt.value());
+}
+
+template <typename DataType>
+enable_if_not_floating_point<DataType, bool> isnan(
+    const util::optional<typename DataType::c_type>& opt) {
+  return false;
+}
+
+template <typename DataType>
+constexpr enable_if_floating_point<DataType, typename DataType::c_type> min() {
+  return -std::numeric_limits<typename DataType::c_type>::infinity();
+}
+
+template <typename DataType>
+constexpr enable_if_floating_point<DataType, typename DataType::c_type> max() {
+  return std::numeric_limits<typename DataType::c_type>::infinity();
+}
+
+template <typename DataType>
+constexpr enable_if_not_floating_point<DataType, typename DataType::c_type> min() {
+  return std::numeric_limits<typename DataType::c_type>::min();
+}
+
+template <typename DataType>
+constexpr enable_if_not_floating_point<DataType, typename DataType::c_type> max() {
+  return std::numeric_limits<typename DataType::c_type>::max();
+}
+
+template <typename DataType>
+Status IsMonotonic(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
```

Comment:
Since this version of `IsMonotonic` requires `DataType` having a `c_type`, this needs to be guarded with [`enable_if_has_c_type`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/type_traits.h#L748). You would need to make other versions for `binary/string` and interval types.